text/0000-low-level-device-model.md

```diff
@@ -0,0 +1,301 @@
+- Feature Name: low_level_device_model
+- Start Date: 2022-07-14
+- RFC PR: [twizzler-rfcs/rfcs#0000](https://github.com/twizzler-operating-system/rfcs/pull/0000)
+- Twizzler Issue: [twizzler-operating-system/twizzler#0000](https://github.com/twizzler-operating-system/twizzler/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC defines the interface that the kernel provides for userspace device management and driver
+implementation. It provides a generic interface that allows for userspace access to
+device memory and features through Twizzler objects, including memory-mapped IO (MMIO) registers,
+interrupt handling, device event messages, and structured device trees.
+
+# Motivation
+[motivation]: #motivation
+
+Devices and drivers are, by and large, very complicated and (along with the surrounding
+infrastructure) make up large parts of modern OS kernels. In a more microkernel fashion, we would
+like to support userspace taking as much responsibility for device management and drivers as
+possible. However, due to the aforementioned complexity, providing a generic interface that allows
+for programming all possible devices is tricky -- we need to provide an interface that allows
+generic access to device memory (and/or IO ports), interrupt progamming services, etc., all while
+trying to minimize the responsibility of in-kernel code to handle these cases.
+
+This RFC does not attempt to describe a higher-level, more convenient driver-programming API set;
+that is planned for a future RFC. Instead, we will cover only the basic user-kernel interface for
+device abstractions.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Twizzler organizes devices and busses into a tree structure, where devices are children of busses,
+which are children of some root node (devices can have children too). These devices are each an
+object that contains information about the device as well as a mechanism of programming device
+interrupts. Each of these devices can also have a number of additional objects attached to it
+(called sub-objects) that provide access to device MMIO registers and bus-specific device
+information.
+
+This low-level interface is not intended to be used by application programmers, and even driver
```

Comment:
I would move this to a BIG BLOCK QUOTE as a warning closer to the top.  Perhaps with HERE THERE BE DRAGONS.